gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,446 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# from abc import ABCMeta, abstractmethod
+from gensim import utils
+import logging
+from timeit import default_timer
+import threading
+from six import string_types
+from collections import defaultdict
+from numpy import vstack
+
+try:
+    from queue import Queue
+except ImportError:
+    from Queue import Queue
+
+logger = logging.getLogger(__name__)
+
+
+# Public Interfaces
+class BaseAny2VecModel(utils.SaveLoad):
+
+    def __init__(self, workers=3, vector_size=100, epochs=5,
+                 callbacks=(), batch_words=10000):
+        """Initialize model parameters.
+        Subclass should initialize the following attributes:
+        - self.kv (instance of concrete implementation of `BaseKeyedVectors` interface)
+        - self.vocabulary (instance of concrete implementation of `BaseVocabBuilder` abstract class)
+        - self.trainables (instance of concrete implementation of `BaseTrainables` abstract clas)
+        """
+        self.vector_size = int(vector_size)
+        self.workers = int(workers)
+        self.epochs = epochs
+        self.train_count = 0
+        self.total_train_time = 0
+        self.batch_words = batch_words
+        self.model_trimmed_post_training = False
+        self.callbacks = callbacks
+
+    def build_vocab(self, data_iterable, update=False, progress_per=10000, **kwargs):
+        """Scan through all the data and create/update vocabulary.
+        Should also initialize/reset/update vectors for new vocab entities.
+        """
+        self.vocabulary.scan_vocab(data_iterable, progress_per=progress_per, **kwargs)
```

Comment:
The problem with this code is that `self.vocabulary` and `self.trainables` don't even exist in the base class. I understand that we rely on the subclass to define these, but it's pretty confusing.

I'm still thinking of a better way of doing this.